src/sql.rs

```diff
@@ -0,0 +1,1175 @@
+use std::collections::HashSet;
+use std::sync::RwLock;
+
+use rusqlite::{Connection, OpenFlags, Statement, NO_PARAMS};
+
+use crate::constants::*;
+use crate::context::Context;
+use crate::dc_param::*;
+use crate::dc_tools::*;
+use crate::error::{Error, Result};
+use crate::peerstate::*;
+use crate::x::*;
+
+const DC_OPEN_READONLY: usize = 0x01;
+
+/// A wrapper around the underlying Sqlite3 object.
+pub struct Sql {
+    pool: RwLock<Option<r2d2::Pool<r2d2_sqlite::SqliteConnectionManager>>>,
+}
+
+impl Sql {
+    pub fn new() -> Sql {
+        Sql {
+            pool: RwLock::new(None),
+        }
+    }
+
+    pub fn is_open(&self) -> bool {
+        self.pool.read().unwrap().is_some()
+    }
+
+    pub fn close(&self, context: &Context) {
+        let mut pool = self.pool.write().unwrap();
+        if pool.is_some() {
```

Comment:
`Option::take()` works the same if there's `None` inside, we can skip this if condition.